src/framework/midi/internal/midiconfiguration.cpp

```diff
@@ -131,97 +130,79 @@ io::path MidiConfiguration::stateFilePath() const
-                            SynthesizerState::ValID id = static_cast<SynthesizerState::ValID>(xml.attributes().value("id").toInt());
-                            g.vals.push_back(SynthesizerState::Val(id, xml.readElementText().toStdString()));
+                while (xml.tokenType() != XmlReader::EndElement) {
+                    if (xml.tokenType() == XmlReader::StartElement) {
+                        if (xml.tagName() == "val") {
+                            SynthesizerState::ValID id = static_cast<SynthesizerState::ValID>(xml.intAttribute("id"));
+                            group.vals.push_back(SynthesizerState::Val(id, xml.readString()));
                         } else {
                             xml.skipCurrentElement();
                         }
                     }
                     xml.readNext();
                 }
 
-                state.groups.insert({ g.name, std::move(g) });
+                state.groups.insert({ group.name, std::move(group) });
             }
         }
     }
 
     if (xml.hasError()) {
-        LOGE() << "failed parse xml, error: " << xml.errorString() << ", path: " << path;
-        return false;
+        LOGE() << "failed parse xml, error: " << xml.error() << ", path: " << path;
     }
 
-    return true;
+    return xml.hasError();
 }
 
 bool MidiConfiguration::writeState(const io::path& path, const SynthesizerState& state)
 {
-    QFile file(path.toQString());
-    if (!file.open(QIODevice::WriteOnly)) {
-        LOGE() << "failed open " << path;
-        return false;
-    }
-
-    QXmlStreamWriter xml(&file);
-    xml.setAutoFormatting(true);
+    XmlWriter xml(path);
     xml.writeStartDocument();
 
     xml.writeStartElement("Synthesizer");
 
     for (auto it = state.groups.cbegin(); it != state.groups.cend(); ++it) {
-        const SynthesizerState::Group& g = it->second;
-
-        if (!g.name.empty()) {
-            xml.writeStartElement(QString::fromStdString(g.name));
-            for (const SynthesizerState::Val& v : g.vals) {
-                xml.writeStartElement("val");
-                xml.writeAttribute("id", QString::number(static_cast<int>(v.id)));
-                xml.writeCharacters(QString::fromStdString(v.val));
-                xml.writeEndElement();
-            }
+        const SynthesizerState::Group& group = it->second;
+
+        if (group.name.empty()) {
+            continue;
+        }
+
+        xml.writeStartElement(group.name);
+        for (const SynthesizerState::Val& value : group.vals) {
+            xml.writeStartElement("val");
+            xml.writeAttribute("id", std::to_string(static_cast<int>(value.id)));
+            xml.writeCharacters(value.val);
             xml.writeEndElement();
         }
+        xml.writeEndElement();
     }
 
     xml.writeEndElement();
     xml.writeEndDocument();
 
     if (xml.hasError()) {
         LOGE() << "failed write xml";
-        return false;
-    }
-
-    if (!file.flush()) {
```

Comment:
should we flush the file in the XmlWriter?

Reply:
QFile does flush() in close()
https://doc.qt.io/qt-5/qfiledevice.html#close